## Shopify Data Science Intern Challenge Solution


#### Question 1: Given some sample data, write a program to answer the following:

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis.

1. Think about what could be going wrong with our calculation. 
2. Think about a better way to evaluate this data. 
3. What metric would you report for this dataset?
4. What is its value?




#### Question 2:

For this question you’ll need to use SQL. Follow this link: https://www.w3schools.com/SQL/TRYSQL.ASP?FILENAME=TRYSQL_SELECT_ALL to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.
1. How many orders were shipped by Speedy Express in total?
2. What is the last name of the employee with the most orders?¶
3. What product was ordered the most by customers in Germany?¶

In [1]:
import numpy as np
import pandas as pd 

import os

In [4]:
pwd

'/root/shared/Downloads'

In [31]:
df = pd.read_csv("/root/shared/Downloads/AOVdataset.csv")
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

## Exploratory Data Analysis

In [3]:
df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

There is a huge difference between the mean(3145.13) and the median(284) which shows that there might be some anamolies in the dataset

In [7]:
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [8]:
df.isna().sum()

order_id          0
shop_id           0
user_id           0
order_amount      0
total_items       0
payment_method    0
created_at        0
dtype: int64

The dataset has no missing values

### Check AOV for each shop

In [10]:
df_group_shop = pd.DataFrame({'avg_amount': df.groupby('shop_id')['order_amount'].mean()}).reset_index()
df_group_shop.sort_values(by=['avg_amount'])

shop_id     avg_amount
91       92     162.857143
1         2     174.327273
31       32     189.976190
99      100     213.675000
52       53     214.117647
..      ...            ...
37       38     390.857143
89       90     403.224490
49       50     403.545455
77       78   49213.043478
41       42  235101.490196

[100 rows x 2 columns]

The average amount for shops 78 and 42 seems very high. There might be an error while populating the data for these 2 shops

In [32]:
df_shop78 = df[df['shop_id']==78]
df_shop78.head(100)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

It appears that the order amount for all the orders of Shop 78 have beeen **multiplied by 100**

In [22]:
df_shop42 = df[df['shop_id']==42]
df_shop42.head(15).sort_values(by=['order_amount'], ascending=False)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
938        939       42      808          1056            3    credit_card   
409        410       42      904           704            2    credit_card   
835        836       42      819           704            2           cash   
40          41       42      793           352            1    credit_card   
308        309       42      770           352            1    credit_card   
834        835       42      792           352            1           cash   
979        980       42      744           352            1          debit   

               created_at  
15     2017-03-07 4:00:00  
60     2017-03-04 4:00:00  
520    2017-03-02 4:00:00  
1104   2017-03-24 4:00:00  
1362   2017-03-15 4:00:00  
1436   2017-03-11 4:00:00  
1364   2017-03-10 6:28:21  
1367   2017-03-13 2:38:34  
938   2017-03-13 23:43:45  
409   2017-03-04 14:32:58  
835   2017-03-09 14:15:15  
40    2017-03-24 14:15:41  
308   2017-03-11 18:14:39  
834   2017-03-25 21:31:25  
979   2017-03-12 13:09:04

The user **607 purchased 2000 items** each time that costed $704000 from the same shop at the same time on different days. This appears to be a fradulent transaction.

## Data Processing

Lets first divide the order amount values by 100 for shop 78 and replace the invalid values

In [33]:
df.loc[df["shop_id"] == 78, "order_amount"] = df["order_amount"]/100

In [34]:
df_shop78_new = df[df['shop_id']==78]
df_shop78_new.head(100)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990        257.25            1    credit_card   
490        491       78      936        514.50            2          debit   
493        494       78      983        514.50            2           cash   
511        512       78      967        514.50            2           cash   
617        618       78      760        514.50            2           cash   
691        692       78      878       1543.50            6          debit   
1056      1057       78      800        257.25            1          debit   
1193      1194       78      944        257.25            1          debit   
1204      1205       78      970        257.25            1    credit_card   
1259      1260       78      775        771.75            3    credit_card   
1384      1385       78      867        257.25            1           cash   
1419      1420       78      912        257.25            1           cash   
1452      1453       78      812        257.25            1    credit_card   
1529      1530       78      810        514.50            2           cash   
2270      2271       78      855        257.25            1    credit_card   
2452      2453       78      709        514.50            2           cash   
2492      2493       78      834       1029.00            4          debit   
2495      2496       78      707        514.50            2           cash   
2512      2513       78      935        514.50            2          debit   
2548      2549       78      861        257.25            1           cash   
2564      2565       78      915        771.75            3          debit   
2690      2691       78      962        771.75            3          debit   
2773      2774       78      890        257.25            1           cash   
2818      2819       78      869        514.50            2          debit   
2821      2822       78      814        514.50            2           cash   
2906      2907       78      817        771.75            3          debit   
2922      2923       78      740        257.25            1          debit   
3085      3086       78      910        257.25            1           cash   
3101      3102       78      855        514.50            2    credit_card   
3151      3152       78      745        257.25            1    credit_card   
3167      3168       78      927        514.50            2           cash   
3403      3404       78      928        771.75            3          debit   
3440      3441       78      982        257.25            1          debit   
3705      3706       78      828        514.50            2    credit_card   
3724      3725       78      766        771.75            3    credit_card   
3780      3781       78      889        257.25            1           cash   
4040      4041       78      852        257.25            1           cash   
4079      4080       78      946        514.50            2           cash   
4192      4193       78      787        771.75            3    credit_card   
4311      4312       78      960        514.50            2          debit   
4412      4413       78      756        514.50            2          debit   
4420      4421       78      969        771.75            3          debit   
4505      4506       78      866        257.25            1          debit   
4584      4585       78      997        257.25            1           cash   
4715      4716       78      818        771.75            3          debit   
4918      4919       78      823        257.25            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

Remove fradulent orders from user 607

In [35]:
df = df[df['user_id']!=607]
df_shop42_607 = df[df['shop_id']==42]
df_shop42_607.head(15).sort_values(by=['order_amount'], ascending=False)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
1364      1365       42      797        1760.0            5           cash   
1367      1368       42      926        1408.0            4           cash   
1471      1472       42      907        1408.0            4          debit   
938        939       42      808        1056.0            3    credit_card   
409        410       42      904         704.0            2    credit_card   
835        836       42      819         704.0            2           cash   
1520      1521       42      756         704.0            2          debit   
1911      1912       42      739         704.0            2           cash   
2003      2004       42      934         704.0            2           cash   
40          41       42      793         352.0            1    credit_card   
308        309       42      770         352.0            1    credit_card   
834        835       42      792         352.0            1           cash   
979        980       42      744         352.0            1          debit   
1512      1513       42      946         352.0            1          debit   
1929      1930       42      770         352.0            1    credit_card   

               created_at  
1364   2017-03-10 6:28:21  
1367   2017-03-13 2:38:34  
1471  2017-03-12 23:00:22  
938   2017-03-13 23:43:45  
409   2017-03-04 14:32:58  
835   2017-03-09 14:15:15  
1520  2017-03-22 13:10:31  
1911   2017-03-07 5:42:52  
2003   2017-03-26 9:21:26  
40    2017-03-24 14:15:41  
308   2017-03-11 18:14:39  
834   2017-03-25 21:31:25  
979   2017-03-12 13:09:04  
1512  2017-03-24 13:35:04  
1929   2017-03-17 8:11:13

In [36]:
df['order_amount'].mean()

304.330323098535

## Thus after removing the fraudulent orders and correcting the calculations in Order amount, we get a **Average order value(AOV) of 304.33** 

## Question 2

### Q1. How many orders were shipped by Speedy Express in total?

<br> SELECT count(OrderID) FROM
<br> Orders o JOIN Shippers s
<br> ON o.ShipperID = s.Shipper|ID
<br> WHERE ShipperName = 'Speedy Express';
<br> <b> Result:</b>
<br> No. of orders for Speedy Express: <b> 54 </b>

### Q2. What is the last name of the employee with the most orders?

<br> SELECT e.LastName, Count(o.OrderID) As NumberOfOrders From Orders o
<br> JOIN Employees e ON o.EmployeeID = e.EmployeeID
<br> Group by LastName Order by NumberOfOrders Desc
<br> Limit 1;
<br> <b> Result:</b>
<br> Last name of employee with most orders(40) is <b> Peacock </b>

### Q3. What product was ordered the most by customers in Germany?

<br> WITH T1 as (SELECT CustomerId FROM Customers WHERE Country='Germany'),
<br> T2 as (SELECT OrderID FROM Orders WHERE CustomerID IN T1),
<br> T3 as (SELECT ProductID, COUNT(OrderID) FROM OrderDetails WHERE OrderID IN T2 
<br>       GROUP BY 1 
<br>       ORDER BY 2 DESC
<br>       LIMIT 1)
<br> 
<br> SELECT ProductName FROM Products WHERE ProductID = (SELECT ProductID FROM T3);

<br> <b> Result:</b>
<br> The product that was ordered the most by customers in Germany is <b> Gorgonzola Telino </b>